compiler/damlc/lib/DA/Cli/Damlc.hs

```diff
@@ -531,72 +532,51 @@ createProjectPackageDb opts fps = do
-                  BSL.writeFile path (ZipArchive.fromEntry dalf)
-              forM_ confFiles $ \conf ->
-                  BSL.writeFile
-                      (dbPath </> "package.conf.d" </>
-                       (takeFileName $ ZipArchive.eRelativePath conf))
-                      (ZipArchive.fromEntry conf)
-              forM_ srcs $ \src -> do
-                  let path = dbPath </> ZipArchive.eRelativePath src
-                  write path (ZipArchive.fromEntry src)
-              ghcPkgPath <- getGhcPkgPath
-              callProcess
-                  (ghcPkgPath </> exe "ghc-pkg")
-                  [ "recache"
-                  -- ghc-pkg insists on using a global package db and will try
-                  -- to find one automatically if we don’t specify it here.
-                  , "--global-package-db=" ++ (dbPath </> "package.conf.d")
-                  , "--expand-pkgroot"
-                  ]
-       | length uniqSdkVersions > 1 && optAllowDifferentSdks opts ->
-            -- when we compile packages with different sdk versions, we need to generate the interface files
+    if | length uniqSdkVersions > 1 && optAllowDifferentSdks opts || not (null fpDalfs) ->
+            -- when we compile packages with different sdk versions or with dalf dependencies, we
+            -- need to generate the interface files
            do loggerH <- getLogger opts "generate interface files"
+              let dalfsFromDars =
+                      [ ( dropExtension $ takeFileName $ ZipArchive.eRelativePath e
+                        , BSL.toStrict $ ZipArchive.fromEntry e)
+                      | e <- dalfs
+                      ]
+              dalfsFromFps <-
+                  forM fpDalfs $ \fp -> do
+                      bs <- B.readFile fp
+                      pure (dropExtension $ takeFileName fp, bs)
+              let allDalfs = dalfsFromDars ++ dalfsFromFps
               pkgs <-
-                  forM dalfs $ \dalf -> do
-                      let dalfBS = BSL.toStrict $ ZipArchive.fromEntry dalf
+                  forM allDalfs $ \(name, dalf) -> do
                       (pkgId, package) <-
                           either (fail . DA.Pretty.renderPretty) pure $
-                          Archive.decodeArchive dalfBS
-                      pure
-                          ( LF.rewriteSelfReferences pkgId package
-                          , stringToUnitId $
-                            dropExtension $
-                            takeFileName $ ZipArchive.eRelativePath dalf)
+                          Archive.decodeArchive dalf
+                      pure (pkgId, package, dalf, stringToUnitId name)
               let pkgMap =
-                      MS.fromList
-                          [ (pkgId, unitId)
-                          | (LF.ExternalPackage pkgId _pkg, unitId) <- pkgs
-                          ]
+                      MS.fromList [(pkgId, unitId) | (pkgId, _pkg, _bs, unitId) <- pkgs]
```

Comment:
I guess deduplication of DALFs with the same package id happens at this step? probably worth pointing out in a comment.

Reply:
added a comment.